### DISCUSSION TITLE
Crashing loop of emqx pod

### DISCUSSION THREAD

**Discussion by and-dzh3** [emqx/emqx#4346] [Q&A]
hello I’m experience a problem with the crashing loop of emqx pod
emqx running in k8s on spot nodes without persistence volume enabled
I experiencing a lot when pod  restarts and cant join to cluster
`xec: /opt/emqx/erts-10.7.1/bin/erlexec -boot /opt/emqx/releases/v4.1.0/emqx -mode embedded -boot_var ERTS_LIB_DIR /opt/emqx/erts-10.7.1/../lib -mnesia dir "/opt/emqx/data/mnesia/emqx@emqx-3.emqx-headless.emqx.svc.cluster.local" -config /opt/emqx/data/configs/app.2021.03.16.07.15.36.config -args_file /opt/emqx/data/configs/vm.2021.03.16.07.15.36.args -vm_args /opt/emqx/data/configs/vm.2021.03.16.07.15.36.args -start_epmd false -epmd_module ekka_epmd -proto_dist ekka -- console
Root: /opt/emqx
/opt/emqx
Starting emqx on node emqx@emqx-3.emqx-headless.emqx.svc.cluster.local
2021-03-16 07:15:39.282 [error] Mnesia('emqx@emqx-3.emqx-headless.emqx.svc.cluster.local'): ** ERROR ** (core dumped to file: "/opt/emqx/MnesiaCore.emqx@emqx-3.emqx-headless.emqx.svc.cluster.local_1615_878939_281573")
** FATAL ** Failed to merge schema: Bad cookie in table definition emqx_retainer: 'emqx@emqx-3.emqx-headless.emqx.svc.cluster.local' = {cstruct,emqx_retainer,set,['emqx@emqx-3.emqx-headless.emqx.svc.cluster.local','emqx@emqx-2.emqx-headless.emqx.svc.cluster.local','emqx@emqx-1.emqx-headless.emqx.svc.cluster.local','emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'],[],[],[],0,read_write,false,[],[],false,retained,[topic,msg,expiry_time],[],[],[{dets,[{auto_save,1000}]},{ets,[compressed,{read_concurrency,true},{write_concurrency,true}]}],{{1615633083813951936,-576460752303423355,1},'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'},{{5,0},{'emqx@emqx-3.emqx-headless.emqx.svc.cluster.local',{1615,727916,663725}}}}, 'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local' = {cstruct,emqx_retainer,set,['emqx@emqx-1.emqx-headless.emqx.svc.cluster.local','emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'],[],[],[],0,read_write,false,[],[],false,retained,[topic,msg,expiry_time],[],[],[{dets,[{auto_save,1000}]},{ets,[compressed,{read_concurrency,true},{write_concurrency,true}]}],{{1615841356561557234,-576460752303422812,1},'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'},{{3,0},{'emqx@emqx-1.emqx-headless.emqx.svc.cluster.local',{1615,841384,819986}}}}`

 EMQX_CLUSTER__K8S__APISERVER: "https://kubernetes.default.svc:443"
  EMQX_CLUSTER__K8S__ADDRESS_TYPE: "hostname"
  EMQX_CLUSTER__K8S__SUFFIX: "svc.cluster.local"
  EMQX_CLUSTER__K8S__APP_NAME: "emqx"
  EMQX_CLUSTER__K8S__SERVICE_NAME: "emqx"
  EMQX_CLUSTER__K8S__NAMESPACE: "emqx"

What can be a root couse ? Bad cookie in table definition emqx_retainer

**Comment by qzhuyan** [comment]
The two cstruct looks like this 
```
#cstruct{
    name = emqx_retainer,type = set,
    ram_copies =
        ['emqx@emqx-3.emqx-headless.emqx.svc.cluster.local',
         'emqx@emqx-2.emqx-headless.emqx.svc.cluster.local',
         'emqx@emqx-1.emqx-headless.emqx.svc.cluster.local',
         'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'],
    disc_copies = [],disc_only_copies = [],external_copies = [],
    load_order = 0,access_mode = read_write,majority = false,
    index = [],snmp = [],local_content = false,
    record_name = retained,
    attributes = [topic,msg,expiry_time],
    user_properties = [],frag_properties = [],
    storage_properties =
        [{dets,[{auto_save,1000}]},
         {ets,
             [compressed,
              {read_concurrency,true},
              {write_concurrency,true}]}],
    cookie =
        {{1615633083813951936,-576460752303423355,1},
         'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'},
    version =
        {{5,0},
         {'emqx@emqx-3.emqx-headless.emqx.svc.cluster.local',
             {1615,727916,663725}}}}


#cstruct{
    name = emqx_retainer,type = set,
    ram_copies =
        ['emqx@emqx-1.emqx-headless.emqx.svc.cluster.local',
         'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'],
    disc_copies = [],disc_only_copies = [],external_copies = [],
    load_order = 0,access_mode = read_write,majority = false,
    index = [],snmp = [],local_content = false,
    record_name = retained,
    attributes = [topic,msg,expiry_time],
    user_properties = [],frag_properties = [],
    storage_properties =
        [{dets,[{auto_save,1000}]},
         {ets,
             [compressed,
              {read_concurrency,true},
              {write_concurrency,true}]}],
    cookie =
        {{1615841356561557234,-576460752303422812,1},
         'emqx@emqx-0.emqx-headless.emqx.svc.cluster.local'},
    version =
        {{3,0},
         {'emqx@emqx-1.emqx-headless.emqx.svc.cluster.local',
             {1615,841384,819986}}}}

```


As you can see the cookie for the same node `emqx@emqx-0.emqx-headless.emqx.svc.cluster.local` looks different.

Could be that the emqx-0 pod losts its state after reset, have you tried with persistence volume enabled?

**Reply by and-dzh3** [reply]
No, if I enable persist volume, what will be effect?
Should I also change something in the broker config?

**Reply by qzhuyan** [reply]
From what I can see two nodes are running different copies of tables in the same cluster thus mnesia refuse to start.

The unique cookie is set while a table is created. So my understanding is there are two `emqx@emqx-0.emqx-headless.emqx.svc.cluster.local` nodes created two copy of tables.

I guess you are running into the following scenario:
1. node 0 starts, node discovery didn't find another node it creates table with cookie A.
2. node 3 starts, node discovery shows there is node 0 online, so it copies the table from node 0 with cookie A. 
3. node 0 pod restart for some reason (quit normal in k8s), node discovery didn't find other node(because node3 isn't fully up yet, no resp to restapi calls) it creates a table with cookie B.
4. node 3 erlang beam restart, node discovery shows there is node 0 online but cookie isn't matching thus restart again.
5. After a few erlang restarts of node 3, k8s should able to kill it since the liveness check should fail.
6. node 3 pod restart with empty data, it should able to find node 0 and copy the data hopefully.

Thus I would like to know if enable PVC would solve the issue at step 3 that if node 0 is up with old data, it shall not recreate the table.

**Reply by and-dzh3** [reply]
I will test and will back with result
